public/app/features/explore/state/utils.ts

```diff
@@ -119,12 +123,55 @@ export async function loadAndInitDatasource(
     }
   }
 
-  const historyKey = `grafana.explore.history.${instance.meta?.id}`;
```

Comment:
Does it mean that existing autocomplete history will not be accessible anymore? Wonder if we should keep reading from both for some time.

Reply:
Added back! The history now will add in query history, query history of mixed queries where the individual query matches the datasource, then local storage query history entries. It will still stop trying to add if we ever hit the max number, and will cap the final result to the max number.